contents/blog/secrets-of-posthog-query-performance.md

```diff
@@ -0,0 +1,152 @@
+
+One of the first issues we saw after moving to ClickHouse was that, for our largest users, filtering by properties was slow.
+
+[Benchmarking these queries using flamegraphs](https://github.com/Slach/clickhouse-flamegraph) showed that the slowness came from two things: reading JSON properties from disk and (to a lesser extent) parsing it during query-time.
+
+We ended up [materializing the most used properties into new columns](/blog/clickhouse-materialized-columns), leveraging the strength of a columnar database without running into column number limits.
+
+Specifically, the new materialized columns are fast to read from disk as they compress really well and ClickHouse can skip parsing JSON entirely during queries.
+
+On our PostHog Cloud setup, we saw this feature improve query performance by 55% on average, with the p99 improvement being 25x.
+
+## Making JOINs 10x faster for large users
+
+Over time, for larger PostHog users with over 10 million visitors, some simple queries like a count of unique users started timing out or running into memory errors.
+
+We narrowed this down to one particular JOIN in our system:
+
+```sql
+JOIN (
+    SELECT distinct_id, argMax(person_id, _timestamp) as person_id
+    FROM (
+        SELECT distinct_id, person_id, max(_timestamp) as _timestamp
+        FROM person_distinct_id
+        WHERE project_id = 123
+        GROUP BY person_id, distinct_id, team_id
+        HAVING max(is_deleted) = 0
+    )
+    GROUP BY distinct_id
+) AS pdi
+
+-- Table schema
+CREATE TABLE person_distinct_id
+(
+    distinct_id VARCHAR,
+    person_id UUID,
+    project_id Int64,
+    _sign Nullable(Int8),
+    is_deleted Nullable(Int8),
+    _timestamp DateTime
+) ENGINE = CollapsingMergeTree(_sign)
+```
+
+This JOIN is as complicated as it is due to a restriction from ClickHouse: updating data is expensive. In ClickHouse, updates require rewriting whole parts of the table instead of individual rows.
+
+Instead, the recommended approach is to use a ReplacingMergeTree or CollapsingMergeTree [TODO LINKS] table engine and handle updating logic at query-time.
+
+During data ingestion, when a given `distinct_id` had its `person_id` changed, PostHog emits a row with `is_deleted=1` for the old `person_id` and a new row with `is_deleted=0`. The above query would then resolve the `distinct_id` => `person_id` mapping at query time.
+
+However, in practice, this query was slow and used up too much of memory, due to needing a subquery to aggregate data correctly. It also had subtle issues with using timestamps for versioning, which was problematic when ClickHouse encountered equal timestamps.
+
+After noticing the problem, we realized we didn't need to actually emit rows with `is_deleted=0` to behave correctly, and could move to an alternative schema, which can be queried as follows:
```

Comment:
Should we explain this change or point out to another resource that explains it?

Reply:
maybe have the new create table also, just for clarity